lib/gosu/window.rb

```diff
@@ -0,0 +1,284 @@
+    attach_function :Gosu_Window_default_button_down, [:pointer, :uint32], :void
+
+    attach_function :Gosu_Window_show, [:pointer], :void
+    attach_function :Gosu_Window_tick, [:pointer], :bool
+    attach_function :Gosu_Window_close_immediately, [:pointer], :void
+
+    attach_function :Gosu_Window_width, [:pointer], :int
+    attach_function :Gosu_Window_set_width, [:pointer, :int], :void
+    attach_function :Gosu_Window_height, [:pointer], :int
+    attach_function :Gosu_Window_set_height, [:pointer, :int], :void
+
+    attach_function :Gosu_Window_mouse_x, [:pointer], :double
+    attach_function :Gosu_Window_set_mouse_x, [:pointer, :double], :void
+    attach_function :Gosu_Window_mouse_y, [:pointer], :double
+    attach_function :Gosu_Window_set_mouse_y, [:pointer, :double], :void
+
+    attach_function :Gosu_Window_caption, [:pointer], :string
+    attach_function :Gosu_Window_set_caption, [:pointer, :string], :void
+    attach_function :Gosu_Window_update_interval, [:pointer], :double
+    attach_function :Gosu_Window_set_update_interval, [:pointer, :double], :void
+    attach_function :Gosu_Window_resize, [:pointer, :int, :int, :bool], :void
+    attach_function :Gosu_Window_is_fullscreen, [:pointer], :bool
+    attach_function :Gosu_Window_is_resizable, [:pointer], :bool
+    attach_function :Gosu_Window_is_borderless, [:pointer], :bool
+    attach_function :Gosu_Window_set_fullscreen, [:pointer, :bool], :void
+    attach_function :Gosu_Window_set_resizable, [:pointer, :bool], :void
+    attach_function :Gosu_Window_set_borderless, [:pointer, :bool], :void
+
+    attach_function :Gosu_Window_sdl_window, [:pointer], :pointer
+
+    attach_function :Gosu_Window_text_input, [:pointer], :pointer
+    attach_function :Gosu_Window_set_text_input, [:pointer, :pointer], :void
+
+    def initialize(width, height, _fullscreen = nil, _update_interval = nil, _resizable = nil, _borderless = nil,
+                   fullscreen: false, update_interval: 16.66666667, resizable: false, borderless: false)
+      fullscreen = _fullscreen if _fullscreen
+      update_interval = _update_interval if _update_interval
+      resizable = _resizable if _resizable
+      borderless = _borderless if _borderless
+
+      window_flags = Gosu.window_flags(fullscreen: fullscreen, resizable: resizable, borderless: borderless)
+
+      __window = Gosu_Window_create(width, height, window_flags, update_interval)
+      Gosu.check_last_error
+      @memory_pointer = FFI::AutoPointer.new(__window, Gosu::Window.method(:Gosu_Window_destroy))
+      @text_input = nil
+
+      {
+        update: [],
+        draw: [],
+        button_down: [:id],
+        button_up: [:id],
+        gamepad_connected: [:id],
+        gamepad_disconnected: [:id],
+        drop: [:filename],
+        needs_redraw?: [],
+        needs_cursor?: [],
+        close: [],
+        gain_focus: [],
+        lose_focus: [],
+      }.each do |callback, args|
```

Comment:
You define Symbol keys but convert them into Strings later on in any use case. Why not using strings in the first place?

Reply:
It looks nicer, I guess. I tried to refactor this and the counterpart in compat.rb (that defines the protected_ methods), but I think I'll move this to a later PR, as long as the tests pass.